components/collections/src/iterator_utils.rs

```diff
@@ -0,0 +1,363 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::codepointtrie::CodePointMapRange;
+use core::ops::RangeInclusive;
+
+/// This is an iterator that turns an iterator over code point ranges
+/// into an iterator producing the complement set of ranges
+pub(crate) struct RangeListIteratorComplementer<I> {
```

Comment:
Yeah I considered that, I think it comes out to having basically the same complicated logic when written as an iterator. It needs to handle [this logic around edge cases](
https://github.com/unicode-org/icu4x/blob/380409ccb5525de92756ee9315fde09931f347d5/components/collections/src/codepointinvlist/builder.rs#L399), PLUS it needs to manually implement the chunking code (since `.chunks()` is for slices only)

(If `array_chunks` were stable that would have been my plan, since then it can just be done with a slightly complicated mix of iterator adapters)

Reply:
I think it boils down to `self.intervals[1..].chunks(2)` with an optional element `chain`ed to the start or the end

Reply:
But the `CodePointMapRange` true/false approach is growing on me. Because then we give you a more useful primitive.

Reply:
> I think it boils down to `self.intervals[1..].chunks(2)` with an optional element `chain`ed to the start or the end

This does not work, because you cannot `chain` and still get a slice. And without a slice you cannot `chunks`.

Like I said, there's a way to do it with `array_chunks`, there is not a way to do it with `chunks()`.

Reply:
We are returning an iterator. You can chain stuff to the start or end of the iterator, and chunks is the whole middle section of the iterator.

Reply:
I think it can be implemented either using iterator adapters or by making an iterator struct that retains a reference to the off-by-one slice.

Reply:
There are ways to do it with `itertools` (new dep), or I could manually write a tuple impl of `array_chunks`, but then we're back on square 1 with manually written iterators. It _might_ be simpler to have that kind of mix but it does not seem sufficiently simpler to me.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a3f66f03823a8fd3cc8f904cf689d90b

or, polyfilling array_chunks is not actually that hard

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=30e3e5a029e67b92d4994a58c7ba44f3

Reply:
Yep, I got it. The older adapter code I had had a small bug that had made it seem like it needed to be more complex; I was able to fix it up.